File: packages/agents/src/cursor.ts
/**
 * Cursor adapter.
 *
 * Path and hook choices are isolated to this module: when Cursor's real on-disk
 * layout is confirmed, only the values below change, never their consumers.
 *
 * - Skills: project `<project>/.cursor/skills/<name>/`, global
 *   `~/.cursor/skills/<name>/` (Cursor keeps per-user state under `~/.cursor`).
 * - Hooks: `json-merge` into `<base>/.cursor/settings.json` under `hooks`.
 *   Cursor configuration is JSON with no comment syntax, so a JSON merge with an
 *   ownership marker is the appropriate strategy.
 */

import type { AgentTarget, HostEnv } from '@skillkeeper/core';
import { makeAdapter } from './makeAdapter.js';
import { baseDir, fsOf, joinPath } from './paths.js';

function cursorDir(target: AgentTarget, env: HostEnv): string {
  return joinPath(baseDir(target, env), '.cursor');
}

export const cursorAdapter = makeAdapter({
  kind: 'cursor',
  skillsRoot: (target, env) => joinPath(cursorDir(target, env), 'skills'),
  availabilityDir: (env) => joinPath(env.homeDir, '.cursor'),
  guidanceFile: async (target, env) => {
    // Prefer an existing legacy .cursorrules; otherwise the modern rules file.
    const base = baseDir(target, env);
    const legacy = joinPath(base, '.cursorrules');
    if (await fsOf(env).exists(legacy)) return legacy;
    return joinPath(base, '.cursor', 'rules', 'skillkeeper.mdc');
  },
  hook: {
    strategy: 'json-merge',
    async resolveTargetFile(target: AgentTarget, env: HostEnv): Promise<string> {
      return joinPath(cursorDir(target, env), 'settings.json');
    },
  },
});

File: crates/skillkeeper-agents/src/cursor.rs
//! Cursor adapter. Ported from `packages/agents/src/cursor.ts`.
//!
//! - Skills: project `<project>/.cursor/skills/<name>/`, global
//!   `~/.cursor/skills/<name>/`.
//! - Hooks: `json-merge` into `<base>/.cursor/settings.json` under `hooks`.

use skillkeeper_core::ports::{HostEnv, PortResult};

use crate::adapter::{make_adapter, AdapterSpec, AgentAdapter, HookCapability};
use crate::model::{AgentKind, AgentTarget, HookStrategy};
use crate::paths::{base_dir, join_path};

fn cursor_dir(target: &AgentTarget, env: &dyn HostEnv) -> PortResult<String> {
    Ok(join_path(&[&base_dir(target, env)?, ".cursor"]))
}

/// Build the Cursor adapter.
pub fn cursor_adapter() -> AgentAdapter {
    make_adapter(AdapterSpec {
        kind: AgentKind::Cursor,
        skills_root: Box::new(|target, env| Ok(join_path(&[&cursor_dir(target, env)?, "skills"]))),
        availability_dir: Box::new(|env| join_path(&[env.home_dir(), ".cursor"])),
        discovery_group: None,
        guidance_file: Box::new(|fs, target, env| {
            // Prefer an existing legacy .cursorrules; otherwise the modern rules file.
            let base = base_dir(target, env)?;
            let legacy = join_path(&[&base, ".cursorrules"]);
            if fs.exists(&legacy)? {
                return Ok(legacy);
            }
            Ok(join_path(&[&base, ".cursor", "rules", "skillkeeper.mdc"]))
        }),
        hook: HookCapability::new(
            HookStrategy::JsonMerge,
            None,
            None,
            Box::new(|target, env| Ok(join_path(&[&cursor_dir(target, env)?, "settings.json"]))),
        ),
    })
}
